api/nodes/permissions.py

```diff
@@ -15,6 +17,34 @@ def has_object_permission(self, request, view, obj):
             return obj.can_edit(auth)
 
 
+class AdminOrPublic(permissions.BasePermission):
+
+    def has_object_permission(self, request, view, obj):
+        assert isinstance(obj, (Node, User)), 'obj must be a Node or User, got {}'.format(obj)
+        auth = get_user_auth(request)
+        node = Node.load(request.parser_context['kwargs']['node_id'])
+        if request.method in permissions.SAFE_METHODS:
+            return node.is_public or node.can_view(auth)
+        else:
+            return node.has_permission(auth.user, osf_permissions.ADMIN)
+
+
+class ContributorDetailPermissions(permissions.BasePermission):
+    """Permissions for contributor detail page."""
+
+    def has_object_permission(self, request, view, obj):
+        assert isinstance(obj, (Node, User)), 'obj must be User or Node, got {}'.format(obj)
+        auth = get_user_auth(request)
+        node = Node.load(request.parser_context['kwargs']['node_id'])
+        user = User.load(request.parser_context['kwargs']['user_id'])
+        if request.method in permissions.SAFE_METHODS:
+            return node.is_public or node.can_view(auth)
+        elif request.method == 'DELETE':
+            return node.has_permission(auth.user, osf_permissions.ADMIN) or auth.user == user
```

Comment:
In the first link, user 2 is trying to remove user 3. "Self removal" would be user 2 removing user 2.

In the second link, we test whether user 1 *can't* remove themselves because they are the only admin. I don't see a test in which user 1 *can* be removed, so long as the project has a second admin as well.